qlib/CsvUtil.qm

```diff
@@ -203,42 +275,137 @@ class CsvHelper {
-                    fdesc += remove f{h};
-                    continue;
+        foreach hash d in (hfDefs.pairIterator()) {
+            if (!d.value.fdesc) {
+                # TODO: can I modify value or is not ?
+                if (d.value.headers && tolwr)
+                    d.value.headers = (map $1.lwr(), d.value.headers);
+
+                d.value.fdesc = ();
+
+                # make a copy of the field description hash
+                hash f = fields;
+                foreach string h in (d.value.headers) {
+                    if (f{h}) {
+                        # add to field description list (and remove from the field description hash)
+                        d.value.fdesc += remove f{h};
+                        continue;
+                    }
+                    # add an entry with type "*string" to the list
+                    d.value.fdesc += ("type": "*string");
                 }
-                # add an entry with type "*string" to the list
-                fdesc += ("type": "*string");
-            }
 
-            if (f)
-                throw "CSVFILEWRITER-FIELD-ERROR", sprintf("field descriptions were given for the following non-existent fields: %y (known header / field names: %y)", f.keys(), headers);
+                # TODO: is it really error ?
+                if (f && !isMultiType())
+                    throw "CSVFILEWRITER-FIELD-ERROR", sprintf("field descriptions were given for the following non-existent fields: %y (known header / field names: %y)", f.keys(), d.value.headers);
+
+                hfDefs{d.key} = d.value;  # TODO: needed ?
+            }
         }
     }
 
     checkType(string key, string value) {
         if (!Types{value})
             throw "CSV-TYPE-ERROR", sprintf("while processing the hash value for option key \"fields\": unknown field type %y for field %y (supported types: %y)", value, key, Types.keys());
     }
+
+    private addHeaders(softstring id, list h) {
+        remove hfDefs{id};
+        if (!h) return;
+        hash d = (
+            "headers": (),
+            "fdesc": (),
+        );
+
+        foreach any c in (h.iterator()) {
+            try {
+                softstring v = c;
+                d.headers += v;
+            }
+            catch (hash ex) {
+                throw "CSVHELPER-ERROR", sprintf("expecting a string value or a value that can be converted to a string for header option; got %y (type %s) in header list position %d instead (full arg: %y)", c, c.type(), $#, h);
+            }
+        }
+        hfDefs{id} = d;
+    }
+
+    private assignHeaders(any h) {
+        switch (h.typeCode()) {
+            case NT_HASH:
+                # multi type lines
+                foreach hash i in (h.pairIterator()) {
+                    addHeaders(i.key, i.value);
+                }
+                break;
+            case NT_LIST:
+                addHeaders(NULL, h);
```

Comment:
Dne 12.5.2016 v 10:37 David Nichols napsal(a):
>
> In qlib/CsvUtil.qm 
> <https://github.com/qorelanguage/qore/pull/850#discussion_r62983274>:
>
> > +                throw "CSVHELPER-ERROR", sprintf("expecting a string value or a value that can be converted to a string for header option; got %y (type %s) in header list position %d instead (full arg: %y)", c, c.type(), $#, h);
> > +            }
> > +        }
> > +        hfDefs{id} = d;
> > +    }
> > +
> > +    private assignHeaders(any h) {
> > +        switch (h.typeCode()) {
> > +            case NT_HASH:
> > +                # multi type lines
> > +                foreach hash i in (h.pairIterator()) {
> > +                    addHeaders(i.key, i.value);
> > +                }
> > +                break;
> > +            case NT_LIST:
> > +                addHeaders(NULL, h);
>
> I think you want to use |NOTHING| instead of |NULL|; |NULL| is an SQL 
> NULL value which is != |NOTHING| (which is like |null| in other 
> languages like java - we can discuss this design issue of Qore BTW if 
> you want)
>
NOTHING vs. NULL is always tricky for me (similar PHP's is_null() vs. 
is_empty() )  ;-/

I had originally NULL used via softstring,

     private addHeaders(softstring id, list h) {
         remove hfDefs{id};
         ...
         hfDefs{id} = d;

When I use NOTHING I think it changes to:

     private addHeaders(*string id, list h) {
         remove hfDefs{id};
         ...
         hfDefs{id} = d;

Seems it works but I using NOTHING as hash key correct ?

Is exists equal to test value is not NOTHING ?

       return curHFDef != NOTHING;
       return exists curHFDef;


> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub 
> <https://github.com/qorelanguage/qore/pull/850/files/9349bd12f3ccae784912a8b7e2ef7a6fdec3f49a#r62983274>
>


-- 
Tomas Mandys
Qore Technologies
Email: tomas.mandys@qoretechnologies.com
Tel: +420-604690589
Skype: tmandys@hotmail.com